Basis/Array.swift

```diff
@@ -0,0 +1,226 @@
+//
+//  Array.swift
+//  Basis
+//
+//  Created by Robert Widmann on 1/11/15.
+//  Copyright (c) 2015 Robert Widmann. All rights reserved.
+//
+
+public enum ArrayMatcher<A> {
+	case Nil
+	case Cons(A, [A])
+}
+
+/// Returns the first element of a non-empty array.
+///
+/// If the provided array is empty, this function throws an exception.
+public func head<A>(l : [A]) -> A {
+	switch match(l) {
+		case .Nil:
+			assert(false, "Cannot take the head of an empty list.")
+		case .Cons(let x, _):
+			return x
+	}
+}
+
+/// Returns an array of all elements but the first in a non-empty array.
+///
+/// If the provided array if empty, this function throws an exception.
+public func tail<A>(l : [A]) -> [A] {
+	switch match(l) {
+		case .Nil:
+			assert(false, "Cannot take the tail of an empty list.")
+		case .Cons(_, let xs):
+			return xs
+	}
+}
+
+public func cons<T>(x : T) -> [T] -> [T] {
+	return { xs in x <| xs }
+}
+
+public func <|<T>(lhs : T, var rhs : [T]) -> [T] {
+	rhs.insert(lhs, atIndex: 0)
+	return rhs
+}
+
```

Comment:
How about `«` and `»`?  :þ

Honestly, though, `|>` and `<|` are correct no matter how you interpret them in this case. `cons` and `snoc` are the proper specializations of the one-handed versions of `Alternative`'s `<|>`.